File: python-code/src/programmers/pre/hash/player_who_did_not_finish.py
''''
    Dictionary: https://dojang.io/mod/page/view.php?id=2307
'''

def solution(participant, completion):

    assert isinstance(participant, list), 'participant is not list'
    assert isinstance(completion, list), 'completion is not list'

    assert 1 <= len(participant) <= 100000, '1 <= participant <= 100000'
    assert len(participant) - 1 == len(completion), 'completion + 1 = participant'
    
    participant_dict = {}
    # participant_dict = dict.fromkeys(participant, 0)
    for p in participant:
        value = participant_dict.get(p, 0)
        participant_dict.update({
            p: 1 if value <= 0 else value + 1
        })

    for c in completion:
        value = participant_dict.get(c)
        participant_dict.update({
            c: value - 1
        })
    
    answer = ''
    for key in participant_dict.keys():
        if participant_dict.get(key) != 0:
            answer = key
            break

    return answer

# "leo"
participant = ["leo", "kiki", "eden"]
completion = ["eden", "kiki"]
print(solution(participant, completion))

# "vinko"
participant = ["marina", "josipa", "nikola", "vinko", "filipa"]
completion = ["josipa", "filipa", "marina", "nikola"]
print(solution(participant, completion))

# "mislav"
participant = ["mislav", "stanko", "mislav", "ana"]
completion = ["stanko", "ana", "mislav"]
print(solution(participant, completion))

File: js-code/src/programmers/pre/hash/player_who_did_not_finish.ts
function solution(participant: any[], completion: any[]) {
  
  if (participant.length < 1 || 100000 < participant.length) throw new Error('1 <= participant <= 100000');
  if (participant.length - 1 !== completion.length) throw new Error('completion + 1 = participant');

  const pMap = new Map<string, number>();
  for (let i = 0, len = participant.length; i < len; i++) {
    const p: string = participant[i],
        c: string = completion[i];
    pMap.set(p, (pMap.get(p) || 0) + 1);
    pMap.set(c, (pMap.get(c) || 0) - 1);
  }

  for (const [k, v] of pMap) {
    if (v > 0)  return k;
  }

  return null;
}

// "leo"
let participant: string[] = ["leo", "kiki", "eden"]
let completion: string[] = ["eden", "kiki"]
console.log(solution(participant, completion))

// "vinko"
participant = ["marina", "josipa", "nikola", "vinko", "filipa"]
completion = ["josipa", "filipa", "marina", "nikola"]
console.log(solution(participant, completion))

// "mislav"
participant = ["mislav", "stanko", "mislav", "ana"]
completion = ["stanko", "ana", "mislav"]
console.log(solution(participant, completion))